Fix support for musl libc

hartwork opened the pull request:
Hi!

First: This is just a draft. I can turn it into something more mergable once the requirements for a mergable version are clear — please let me know.

---

It [was brought to my attention](https://bugs.gentoo.org/713646) that dcfldd makes use of header `error.h` and function `error(3)` that both glibc has and musl doesn't. My patches here replace these uses and hence fix support for musl. They are based on (1) an idea from https://gitlab.alpinelinux.org/hason/aports/-/blob/v3.3.0/testing/dcfldd/dcfldd-error.patch and (2) reading the `error(3)` man page.

From the `error(3)` man page I would like to quote these highlights:
- "It flushes stdout [..]"
- "[..] and then outputs to stderr the program name, a colon and a space [..]"
- "The output is terminated by a newline character."
- "If status has a nonzero value, then error() calls exit(3) to terminate the program[.]"

More context can be found below:

```
error(3)               Library Functions Manual              error(3)

NAME
       error, error_at_line, error_message_count, error_one_per_line,
       error_print_progname - glibc error reporting functions

LIBRARY
       Standard C library (libc, -lc)

SYNOPSIS
       #include <error.h>

       void error(int status, int errnum, const char *format, ...);
[..]

DESCRIPTION
       error()  is  a  general  error-reporting function.  It flushes
       stdout, and then outputs to stderr the program name,  a  colon
       and a space, the message specified by the printf(3)-style for‐
       mat  string  format, and, if errnum is nonzero, a second colon
       and a space followed by the string given by  strerror(errnum).
       Any  arguments required for format should follow format in the
       argument list.  The output is terminated by a newline  charac‐
       ter.

       The program name printed by error() is the value of the global
       variable  program_invocation_name(3).  program_invocation_name
       initially has the same value as main()’s argv[0].   The  value
       of  this  variable can be modified to change the output of er‐
       ror().

       If status has a nonzero value, then error() calls  exit(3)  to
       terminate  the  program using the given value as the exit sta‐
       tus; otherwise it returns after printing the error message.

       [..]
```

What do you think?

hartwork committed:
util.c: Resolve use of error.h and error(3) to support MUSL

hartwork committed:
xstrtol.{c,h}: Resolve use of error.h and error(3) to support MUSL

davidpolverari commented:
Hi there!

Hmm. Yes, it is bad to be depending on non-standard code like this. I will write a small review of your changes with some of my thoughts.

davidpolverari reviewed (commented):
In general, the code LGTM. Some problems I see are:

- although there are few `error()` calls, the replacement code is repeating itself. Thus, I think we should strive to encapsulate it in an auxiliary replacement function, even though it might sound overkill. 
- the other issue is related to the above: the replacement code uses the hardcoded program name. It is not something really important for the four places the function is invoked from, but it creates inconsistencies (and more work) when changes are made (or a fork, or the user renames the executable...).

davidpolverari commented:
I drafted a replacement on the [dcfldd-error](https://github.com/resurrecting-open-source-projects/dcfldd/tree/dcfldd-error). Of course it has to be reviewed and enhanced, but I think we could use it as a basis to replace the dependency on glibc's `error.h`. What do you think about it?

hartwork commented:
Hi @davidpolverari,


thanks for the interest, the review, the alternative draft of yours!

I agree with your evaluation with regard to duplication and the hardcoded program name.  (I just wanted to be sure of upstream interest before investing more time.)

I would love to do line-based review of your branch if you create a pull request for it.
While we do not have that, I can say that:

- The general direction looks clean and great :+1: :+1: 

- Either of us should add "compiler" `CC=musl-gcc` to the CI matrix so that have practical proof this is fully working at merge time and does not regress in the future:
  - it's package `musl-tools` in Ubuntu and
  - my work on ttyplot included https://github.com/tenox7/ttyplot/pull/165/ for a full demo.

- Use of function `basename` is portability trouble, maybe we should "vendor" `basename` also. My earlier pull request https://github.com/USBGuard/usbguard/pull/514/commits/e023f55f949b09e4161e489b7e65339fffa4fbcc for USBGuard demonstrates the extent of the problem.

- I would persoanlly always use different names for something that is not the original, e.g. not use `error.h` and `error` for a function name. Maybe `libc_error.c` or `error_remake.c` or `port_error.c` or or…

- Current function `error` would probably crash when `program_invocation_name` has been failed to initialize.  It would be nice to not have that.

- Potentially out of scope: there is `program_name = argv[0];` a few lines up, maybe that and `program_invocation_name = basename(argv[0]);` could/should be joined somehow.

- Potentially out of scope: there is also use of `PROGRAM_NAME` that could/should use the dynamic program name

- I would always advise to add copyright headers to new `.c` and `.h` files.

How do we continue? I see at least three options:

- a) you create a pull request and I take the review seat
- b) I replace my pull request content with your commit and build on top of it and you take the review seat
- c) (we take things to a voice call in English or German)
- d) …

What do you think?

Best, Sebastian

davidpolverari commented:
> Hi @davidpolverari,
> 
> thanks for the interest, the review, the alternative draft of yours!
>
Hi there! I'm the one to thank you for the help that you always give to this project!

> I agree with your evaluation with regard to duplication and the hardcoded program name. (I just wanted to be sure of upstream interest before investing more time.)
>
No problems. You told it was just a draft. I only told that to give feedback about things I considered important for us not to forget when fixing the problem.

About your feedback:

> I would love to do line-based review of your branch if you create a pull request for it. While we do not have that, I can say that:
> 
>     * The general direction looks clean and great 👍 👍
> 
Thanks!

>     * Either of us should add "compiler" `CC=musl-gcc` to the CI matrix so that have practical proof this is fully working at merge time and does not regress in the future:
>       
>       * it's package `musl-tools` in Ubuntu and
>       * my work on ttyplot included [Fix compilation with musl libc (fixes #164) + make CI cover compilation with musl libc + bump version to 1.6.2 tenox7/ttyplot#165](https://github.com/tenox7/ttyplot/pull/165) for a full demo.
>
I'd prefer that you make those changes to the CI matrix after I make a PR. You clearly have more experience than me on that front :)

>     * Use of function `basename` is portability trouble, maybe we should "vendor" `basename` also. My earlier pull request [USBGuard/usbguard@e023f55](https://github.com/USBGuard/usbguard/commit/e023f55f949b09e4161e489b7e65339fffa4fbcc) for USBGuard demonstrates the extent of the problem.
>
In hindsight, after reading the manpage for `error(3)`, I see that I used what is the equivalent to `program_invocation_short_name` (which is the basename for `argv[0]`), while `error()` uses `program_invocation_name` (which has the same value as `argv[0]`). Thus, we can remove `basename()` usage altogether.

>     * I would persoanlly always use different names for something that is not the original, e.g. not use `error.h` and `error` for a function name. Maybe `libc_error.c` or `error_remake.c` or `port_error.c` or or…
>
You're right. I'm thinking of using something like `dcfldd_error()` and rename the C and header files to `dcfldd_error.{c,h}` as well. Better names are welcome. :)

>     * Current function `error` would probably crash when `program_invocation_name` has been failed to initialize.  It would be nice to not have that.
> 
Hmm, I can't visualize a scenario in which the program is run and we don't have at least `argv[0]` set. That's why I didn't think about any checks in this case. But feel free to correct me if I missed something.

>     * Potentially out of scope: there is `program_name = argv[0];` a few lines up, maybe that and `program_invocation_name = basename(argv[0]);` could/should be joined somehow.
>
As we are going to remove the usage of `basename()`, the last line will vanish.
 
>     * Potentially out of scope: there is also use of `PROGRAM_NAME` that could/should use the dynamic program name
> 
That would not really be the dynamic program name, as it will always give the same string, independently from the current executable name (`#define PROGRAM_NAME "dcfldd"`). But maybe we could use it, as we don't need to fully mimic glibc `error()`'s behavior, although I'd prefer having it dynamically.

>     * I would always advise to add copyright headers to new `.c` and `.h` files.
> 
Sure, I will add them.

> 
> How do we continue? I see at least three options:
> 
>     * a) you create a pull request and I take the review seat
> 
>     * b) I replace my pull request content with your commit and build on top of it and you take the review seat
> 
>     * c) (we take things to a voice call in English or German)
> 
>     * d) …
> 
> 
> What do you think?
> 

~I think I should create a new PR with the changes discussed here and you could take the review seat. Then we make enhancements from that point~. To avoid creating a new PR, maybe you could replace your pull request with my commit and we build on top of that. I would then make the changes and you would take the review seat. Is that ok?
 
> Best, Sebastian

Wish the same for you!

hartwork committed:
full-check.yml: Cover compilation against musl libc

hartwork committed:
src/sha2.h: Replace SHA2_USE_INTTYPES_H by HAVE_INTTYPES_H for musl libc

Symptom was:
> musl-gcc -DHAVE_CONFIG_H -I. -I..    -fgnu89-inline -std=c99 -Wall -Wextra -pedantic -MT md5.o -MD -MP -MF .deps/md5.Tpo -c -o md5.o md5.c
> In file included from hash.h:31,
>                  from md5.h:49,
>                  from md5.c:35:
> sha2.h:76:9: error: unknown type name ‘u_int8_t’
>    76 | typedef u_int8_t  sha2_byte;    /* Exactly 1 byte */
>       |         ^~~~~~~~

hartwork committed:
full-check.yml: Dump config.log on configure failure for debugging

hartwork committed:
full-check.yml: Exclude sanitizers from musl-gcc

hartwork commented:
> > Hi @davidpolverari,
> > thanks for the interest, the review, the alternative draft of yours!
> 
> Hi there! I'm the one to thank you for the help that you always give to this project!

@davidpolverari glad to hear, glad to help and collaborate with you.

(I remember a colleague showed dcfldd to me with proper progress display long(?) before GNU coreutils 8.24 eventually added `status=progress` to `dd` to stop the `USR1` kill loop annoyance.  I think `dd` is missing `sizeprobe` to this day.  So somehow dcfldd stuck with me :smiley:  How about you?)


> > * Either of us should add "compiler" `CC=musl-gcc` to the CI matrix so that have practical proof this is fully working at merge time and does not regress in the future:
> >   
> >   * it's package `musl-tools` in Ubuntu and
> >   * my work on ttyplot included [Fix compilation with musl libc (fixes #164) + make CI cover compilation with musl libc + bump version to 1.6.2 tenox7/ttyplot#165](https://github.com/tenox7/ttyplot/pull/165) for a full demo.
> 
> I'd prefer that you make those changes to the CI matrix after I make a PR. You clearly have more experience than me on that front :)

I have added four commits now for you to this pull request here to make musl-gcc work fully work in CI.  Commit `src/sha2.h: Replace SHA2_USE_INTTYPES_H by HAVE_INTTYPES_H for musl libc` sort of proves that musl CI and musl fixes cannot really go alone or we'd be blind.  Does cherry-picking those four commits work for you?  I can adjust them as needed.


> > * Use of function `basename` is portability trouble, maybe we should "vendor" `basename` also. My earlier pull request [USBGuard/usbguard@e023f55](https://github.com/USBGuard/usbguard/commit/e023f55f949b09e4161e489b7e65339fffa4fbcc) for USBGuard demonstrates the extent of the problem.
> 
> In hindsight, after reading the manpage for `error(3)`, I see that I used what is the equivalent to `program_invocation_short_name` (which is the basename for `argv[0]`), while `error()` uses `program_invocation_name` (which has the same value as `argv[0]`). Thus, we can remove `basename()` usage altogether.

Interesting catch, that's great news!  PS: I confirm:

```console
$ cat main.c 
#include <errno.h>
#include <error.h>

int main() {
    error(1, ENOMEM, "hello world");
    return 0;
}

$ gcc main.c && "${PWD}"/a.out ; echo $?
/tmp/tmp.KG0dryxNaz/a.out: hello world: Cannot allocate memory
1
```

> > * I would persoanlly always use different names for something that is not the original, e.g. not use `error.h` and `error` for a function name. Maybe `libc_error.c` or `error_remake.c` or `port_error.c` or or…
> 
> You're right. I'm thinking of using something like `dcfldd_error()` and rename the C and header files to `dcfldd_error.{c,h}` as well. Better names are welcome. :)

Good enough for a start, I would say.

> > * Current function `error` would probably crash when `program_invocation_name` has been failed to initialize.  It would be nice to not have that.
> 
> Hmm, I can't visualize a scenario in which the program is run and we don't have at least `argv[0]` set. That's why I didn't think about any checks in this case. But feel free to correct me if I missed something.

If we initialize the global to `NULL` and add a `NULL` check to the function body, it would feel slightly less brittle, but just my 2 cents, I don't have a cleaner solution in mind yet.  I guess it's due to not being able to pull `argv[0]` out of thin air (if that makes sense to say) from anywhere.


> > * Potentially out of scope: there is `program_name = argv[0];` a few lines up, maybe that and `program_invocation_name = basename(argv[0]);` could/should be joined somehow.
> 
> As we are going to remove the usage of `basename()`, the last line will vanish.

Nice!


> 
> > * Potentially out of scope: there is also use of `PROGRAM_NAME` that could/should use the dynamic program name
> 
> That would not really be the dynamic program name, as it will always give the same string, independently from the current executable name (`#define PROGRAM_NAME "dcfldd"`). But maybe we could use it, as we don't need to fully mimic glibc `error()`'s behavior, although I'd prefer having it dynamically.

Understood, yes.  I was bringing it up as potnetially out of scope but related since we may want to replace constant `PROGRAM_NAME` by something based on `argv[0]` also, but I haven't checked its use to the last detail, and its not urgent or a dependency.


> > * I would always advise to add copyright headers to new `.c` and `.h` files.
> 
> Sure, I will add them.

Thank you! :pray:


> > How do we continue? I see at least three options:
> >
> > * a) you create a pull request and I take the review seat
> > 
> > * b) I replace my pull request content with your commit and build on top of it and you take the review seat
> > 
> > * c) (we take things to a voice call in English or German)
> > 
> > * d) …
> >   
> > What do you think?
> 
> ~I think I should create a new PR with the changes discussed here and you could take the review seat. Then we make enhancements from that point~. To avoid creating a new PR, maybe you could replace your pull request with my commit and we build on top of that. I would then make the changes and you would take the review seat. Is that ok?

I think GitHub started allowing to do comment-review on one' own pull but not to request changes or approve…

![review](https://github.com/user-attachments/assets/40dc32d9-c24b-448a-ba33-0d1db6e673b4)

… and — even more importantly — the branch would still be from my fork repository so you could not just push to it.  So I think this very combination will be trouble, technically.  Do you want to avoid the future zombie pull request that #34 would become with (a), primarily?

davidpolverari commented:
Hi!

> (I remember a colleague showed dcfldd to me with proper progress display long(?) before GNU coreutils 8.24 eventually added `status=progress` to `dd` to stop the `USR1` kill loop annoyance. I think `dd` is missing `sizeprobe` to this day. So somehow dcfldd stuck with me 😃 How about you?)
> 
For me, it was about also proper progress display and on-the-fly hashing support.

> I have added four commits now for you to this pull request here to make musl-gcc work fully work in CI. Commit `src/sha2.h: Replace SHA2_USE_INTTYPES_H by HAVE_INTTYPES_H for musl libc` sort of proves that musl CI and musl fixes cannot really go alone or we'd be blind. Does cherry-picking those four commits work for you? I can adjust them as needed.
>
I will cherry-pick them in the process of preparing my PR, thanks a lot!
 
> Interesting catch, that's great news! PS: I confirm:
> 
> ```
> $ cat main.c 
> #include <errno.h>
> #include <error.h>
> 
> int main() {
>     error(1, ENOMEM, "hello world");
>     return 0;
> }
> 
> $ gcc main.c && "${PWD}"/a.out ; echo $?
> /tmp/tmp.KG0dryxNaz/a.out: hello world: Cannot allocate memory
> 1
> ```
>
Thanks for the confirmation!
 
> If we initialize the global to `NULL` and add a `NULL` check to the function body, it would feel slightly less brittle, but just my 2 cents, I don't have a cleaner solution in mind yet. I guess it's due to not being able to pull `argv[0]` out of thin air (if that makes sense to say) from anywhere.
>
Hmm, this would be really better. I only think that we must put a warning on the source that we are not implementing a totally glibc-compliant replacement, and we might be good to go with those differences (eg.: using dcfldd's `program_name`  instead of `program_invocation_name`). BTW, if we end up using dcfldd's own `program_name`, I think the NULL check could be elided, as the function will be used with no compatibility expectations, and we could just use an assert to avoid assuming too much.

In the end, after we drafted a PR, we can decide which level checking we want on that.
 
> I think GitHub started allowing to do comment-review on one' own pull but not to request changes or approve…
> 
> ![review](https://private-user-images.githubusercontent.com/1577132/377062523-40dc32d9-c24b-448a-ba33-0d1db6e673b4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjkxNzk1NzAsIm5iZiI6MTcyOTE3OTI3MCwicGF0aCI6Ii8xNTc3MTMyLzM3NzA2MjUyMy00MGRjMzJkOS1jMjRiLTQ0OGEtYmEzMy0wZDFkYjZlNjczYjQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAxNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMTdUMTUzNDMwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YzdiYzE5MWZkODQ5MmI1OGJjYmRiODcyMjBiNTcxYjc0YjM2Yzc2Njg2ZmM1ZGEyMmEzNzg3ODFhZjY1NTM4ZiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.jTVt1yvDYiLxJ9w5Af2ou8mz6c54ii1REwRna46ZPrg)
> 
> … and — even more importantly — the branch would still be from my fork repository so you could not just push to it. So I think this very combination will be trouble, technically. Do you want to avoid the future zombie pull request that #34 would become with (a), primarily?

Yes, I will go with option (a) then and create the PR. I think I will be able to do it this weekend.

hartwork commented:
> Yes, I will go with option (a) then and create the PR. I think I will be able to do it this weekend.

@davidpolverari perfect, thank you! :pray:

davidpolverari commented:
Could you allow for edits on your PR? I was intending to force-push a branch which contained:

- your musl support commits;
- my dcfld_error module;
- replacement of `error()` by `dcfldd_error()`. 

In this branch, I dropped the first draft of yours, as it was replaced by the `dcfldd_error` implementation. You can see what I mean to force-push on the [hartwork-musl branch](https://github.com/resurrecting-open-source-projects/dcfldd/commits/hartwork-musl/).

davidpolverari commented:
Please, have a look at https://github.com/resurrecting-open-source-projects/dcfldd/pull/37.

hartwork commented:
> Please, have a look at #37.

@davidpolverari lovely, thank you!

hartwork commented:
Closing in favor of #37 …
